### ISSUE TITLE
Monsters and Players need collision detection

### ISSUE THREAD

**Issue by sturmE** [4950/DirtyGame#38]
Right now, we just check if player collides with monster. We need just prevent collisions by setting entitys' positions when they collide.

**Comment by JaredPoetter** [comment]
Are we going to make the boundary box around the player dynamic so that if the player is attacking the box is bigger in the given direction? In the player animation that I have been using has an offset for the sprite animation of the sword swinging. So you will probably need to include this value in the calculation too. There is an offset for all of them but the ones that do not need an offset is set to 0.

**Comment by sturmE** [comment]
Ya we could do that. Right now the bounding box is just hardcoded but it should resize based on the size of the entity.

**Comment by JaredPoetter** [comment]
Should we create a SizeComponent instead of sending; the AnimationComponent to the CollisionSystem, putting the size and offset in the SpatialCompnent and sending it to the AnimationComponent.

The SizeCompnent will have to have the size and the offset from the center to the center of the sprite character itself.

**Comment by sturmE** [comment]
I like the idea of having size in the Spatial Component. Because the purpose of spatial component is to represent the the entity in our simulation. To prevent you from sending animation to the collision system, I think it would be fine to modify the spatial component in your animation system. but not include the spatial component in your animation system's aspect. This would be like says, "k, I'm going to modify the animation. I know it is going to change the size of the entity. Let me check if this entity cares about its size."

So just.

```
//modify animation crap
//notice we changed size of sprite
if(e.HasComponent<Spatial>()) {
    //modify it
}
```

**Comment by JaredPoetter** [comment]
Ok I will do that then. I am assuming that you think that the offset values should be in the spatial component too?

**Comment by sturmE** [comment]
Could you rephrase your description of the offset. I dont entirely understand what its purpose is. That being said, I feel like the offset is an animation thing. I was thinking that if an entity cared about its size (SpatialComponent), your animation system would perform this logic to resize the entities spatial component (the position/width/height) and nothing else.

**Comment by JaredPoetter** [comment]
Well if you look at the player sprite sheet image you can see that the CENTER OF THE PLAYER in the walking player frames are in the center of those. But if you look at the attacking player the CENTER OF THE PLAYER in those frames are NOT in the center of the frame. So the animation needs to render the sprite offset some amount. I am just guessing that to get the collisions to work perfectly we would need to include this offset.

Does this make sense?

**Comment by sturmE** [comment]
Ok I get it. Ya, I having the offset in the spatial component doesnt make sense because we store the center of the player in there already. Adding an offset parameter to the spatial component is unnecessary when you could have just offset his position. If you can think of a way to apply the offset to the spatial component by not having it in the spatial component that'd be better. I think the more challenging part would be to "unapply" the offset. 

If you can think of a great way, just throw it in the spatial component and make a ticket to come back to it.

**Comment by bkern24** [comment]
Just To update what I have been researching the last few days, and my plan for collision detection. 

I am to going first implement the quadtree system simply using rectangles (Not dynamic) and then I my plan was to look at the sprites and their animations to decide what type of shape/object I want to use for each, or a system that can adapt.

I currently am looking into how to implement quad trees without using a library as I want to understand the baseline meaning of collision detection. I then am going to look into using Farseer which allows for both quad and dynamic tree collision detection.

**Comment by bkern24** [comment]
Also I am curious as to what should be occurring when players and monsters collide. I mean if they collide should they get pushed back (For example, in Zelda the player gets hits and slides back a few "steps"). However, I feel as if this may be a different system in the framework??

**Comment by sturmE** [comment]
Whenever a collision happens we want to do something. The two somethings that I can think of are bump into eachother (monsters dont walk on top of eachother) and do damage (sword hits monster, monster takes damage). The collision system could just check if stuff collides, then we need to define what is notified when certain types of things collide.

**Comment by baylorw** [comment]
A suggestion - leave the bounding box for the sprite alone. When attacking, and assuming you know facingDirection, spawn a new entity, Weapon. It has a bounding box at least partially outside the player's box. Doesn't need an image because the animation has that. Monster collides with either the player or the player's weapon and diff things happen depending on what he hit. Think of it like Mario's shoes - when you jump down on a monster it dies, when you run sideways you die.

**Comment by AOEChamp** [comment]
Collision has been handled with physics. Closing
